src/main/scala/devices/i2c/I2C.scala

```diff
@@ -0,0 +1,515 @@
+          }
+
+          load        := true.B
+        }
+      }
+      is (s_byte_start) {
+        when (bitCmdAck) {
+          when (cmd.read) {
+            byteState := s_byte_read
+            bitCmd    := I2C_CMD_READ
+          }
+          .otherwise {
+            byteState := s_byte_write
+            bitCmd    := I2C_CMD_WRITE
+          }
+
+          load        := true.B
+        }
+      }
+      is (s_byte_write) {
+        when (bitCmdAck) {
+          when (bitCntDone) {
+            byteState := s_byte_ack
+            bitCmd    := I2C_CMD_READ
+          }
+          .otherwise {
+            byteState := s_byte_write
+            bitCmd    := I2C_CMD_WRITE
+            shift     := true.B
+          }
+        }
+      }
+      is (s_byte_read) {
+        when (bitCmdAck) {
+          when (bitCntDone) {
+            byteState := s_byte_ack
+            bitCmd    := I2C_CMD_WRITE
+          }
+          .otherwise {
+            byteState := s_byte_read
+            bitCmd    := I2C_CMD_READ
+          }
+
+          shift       := true.B
+          transmitBit := cmd.ack
+        }
+      }
+      is (s_byte_ack) {
+        when (bitCmdAck) {
+          when (cmd.stop) {
+            byteState := s_byte_stop
+            bitCmd    := I2C_CMD_STOP
+          }
+          .otherwise {
+            byteState := s_byte_idle
+            bitCmd    := I2C_CMD_NOP
+
+	    // generate command acknowledge signal
+            cmdAck    := true.B
+          }
+
+	  // assign ack_out output to bit_controller_rxd (contains last received bit)
+          receivedAck := receivedBit
+
+          transmitBit := true.B
+        }
+        .otherwise {
+          transmitBit := cmd.ack
+        }
+      }
+      is (s_byte_stop) {
+        when (bitCmdAck) {
+          byteState := s_byte_idle
+          bitCmd    := I2C_CMD_NOP
+
+	  // assign ack_out output to bit_controller_rxd (contains last received bit)
+          cmdAck    := true.B
+        }
+      }
+    }
+  }
+
+
+  //////// Top level ////////
+
+  // hack
+  val nextCmd = Wire(UInt(8.W))
+  nextCmd := cmd.asUInt
+  cmd := (new CommandBundle).fromBits(nextCmd)
+
+  when (cmdAck || arbLost) {
+    cmd.start := false.B    // clear command bits when done
+    cmd.stop  := false.B    // or when aribitration lost
+    cmd.read  := false.B
+    cmd.write := false.B
+  }
+  cmd.irqAck  := false.B    // clear IRQ_ACK bit (essentially 1 cycle pulse b/c it is overwritten by regmap below)
+
+  status.receivedAck := receivedAck
+  when (stopCond) {
+    status.busy             := false.B
+  }
+  .elsewhen (startCond) {
+    status.busy             := true.B
+  }
+
+  when (arbLost) {
+    status.arbLost          := true.B
+  }
+  .elsewhen (cmd.start) {
+    status.arbLost          := false.B
+  }
+  status.transferInProgress := cmd.read || cmd.write
+  status.irqFlag            := (cmdAck || arbLost || status.irqFlag) && !cmd.irqAck
+
+  // Note that these are out of order.
```

Comment:
Because at some point I re-ordered the GPIO CSRs in the spec and didn't feel like re-ordering the code. If you look at the defined offsets, they are out of order.